lib/src/dartdoc_options.dart

```diff
@@ -9,17 +9,707 @@
+  String get fieldName => keys.join('.');
+
+  @override
+  void _onMissing(
+      _OptionValueWithContext valueWithContext, String missingPath) {
+    String dartdocYaml = pathLib.join(
+        valueWithContext.canonicalDirectoryPath, valueWithContext.definingFile);
+    throw new DartdocFileMissing(
+        'Field ${fieldName} from ${dartdocYaml}, set to ${valueWithContext.value}, resolves to missing path: "${missingPath}"');
+  }
+
+  @override
+
+  /// Searches for a value in configuration files relative to [dir], and if not
+  /// found, returns [defaultsTo].
+  T valueAt(Directory dir) {
+    return _valueAtFromFiles(dir) ?? defaultsTo;
+  }
+
+  T _valueAtFromFiles(Directory dir) {
+    _OptionValueWithContext valueWithContext;
+    if (parentDirOverridesChild) {
+      valueWithContext = _valueAtFromFilesLastFound(dir);
+    } else {
+      valueWithContext = _valueAtFromFilesFirstFound(dir);
+    }
+    return _handlePathsInContext(valueWithContext);
+  }
+
+  /// Searches all dartdoc_options files through parent directories,
+  /// starting at [dir], for the option and returns one once
+  /// found.
+  _OptionValueWithContext _valueAtFromFilesFirstFound(Directory dir) {
+    _OptionValueWithContext value;
+    while (true) {
+      value = _valueAtFromFile(dir);
+      if (value != null || pathLib.equals(dir.parent.path, dir.path)) break;
+      dir = dir.parent;
+    }
+    return value;
+  }
+
+  /// Searches all dartdoc_options files for the option, and returns the
+  /// value in the top-most parent directory dartdoc_options.yaml file it is
+  /// mentioned in.
+  _OptionValueWithContext _valueAtFromFilesLastFound(Directory dir) {
+    _OptionValueWithContext value;
+    while (true) {
+      _OptionValueWithContext tmpValue = _valueAtFromFile(dir);
+      if (tmpValue != null) value = tmpValue;
+      if (pathLib.equals(dir.parent.path, dir.path)) break;
+      dir = dir.parent;
+    }
+    return value;
+  }
+
+  /// Returns null if not set in the yaml file in this directory (or its
+  /// parents).
+  _OptionValueWithContext _valueAtFromFile(Directory dir) {
+    _YamlFileData yamlFileData = _yamlAtDirectory(dir);
+    String contextPath = yamlFileData.canonicalDirectoryPath;
+    var yamlData = yamlFileData.data;
+    for (String key in keys) {
+      if (!yamlData.containsKey(key)) return null;
+      yamlData = yamlData[key];
+    }
+
+    var returnData;
+    if (_isListString) {
+      if (yamlData is YamlList) {
+        returnData = <String>[];
+        for (var item in yamlData as YamlList) {
+          returnData.add(item.toString());
+        }
+      }
+    } else if (_isMapString) {
+      if (yamlData is YamlMap) {
+        returnData = <String, String>{};
+        for (MapEntry entry in yamlData.entries) {
+          returnData[entry.key.toString()] = entry.value.toString();
+        }
+      }
+    } else if (_isDouble) {
+      if (yamlData is num) {
+        returnData = (yamlData as num).toDouble();
```

Comment:
I'm sort of surprised this cast is necessary after the `is` check...

Reply:
The analyzer seems to think yamlData is a Map here, which should be impossible.
https://github.com/dart-lang/sdk/issues/32880

Redid this to explicitly specify the variable yamlData as dynamic, then type inference works as wexpected.